File: backend/app/models/trade.py
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.emotion import EmotionResponse
from app.models.tag import TagResponse


class ActivityType(str, Enum):
  buy = "Buy"
  sell = "Sell"


class TradeDirection(str, Enum):
  long = "Long"
  short = "Short"


class TradeStatus(str, Enum):
  open = "Open"
  closed = "Closed"
  partial = "Partial"
  breakeven = "Breakeven"


class AccountType(str, Enum):
  test = "test"
  demo = "demo"
  live = "live"


class TradeActivityCreate(BaseModel):
  """A single buy or sell execution to attach to a trade."""

  model_config = ConfigDict(use_enum_values=True)

  type: ActivityType
  price: float = Field(gt=0)
  quantity: float = Field(gt=0)
  date: str = Field(min_length=1)


class TradeActivityResponse(BaseModel):
  """Response schema for a single trade activity."""

  id: int
  trade_id: int
  type: ActivityType
  price: float
  quantity: float
  date: str
  is_entry: bool


class TradeScreenshotResponse(BaseModel):
  """Response schema for a trade screenshot record."""

  id: int
  trade_id: int
  filename: str
  timeframe_unit: Optional[str] = None
  timeframe_value: Optional[int] = None
  label: Optional[str] = None
  created_at: Optional[str] = None


class TradeCreate(BaseModel):
  """Request body for creating a trade."""

  model_config = ConfigDict(use_enum_values=True)

  asset_id: int
  account_type: AccountType = AccountType.live
  stop_loss: Optional[float] = None
  notes: Optional[str] = Field(default=None, max_length=2000)
  missed_opportunity: bool = False
  risk_per_trade: Optional[float] = None
  timeframe_unit: Optional[str] = Field(default=None, max_length=20)
  timeframe_value: Optional[int] = None
  # Per-trade chart timeframe overrides (issue #236). None falls back to the
  # user's chart_timeframes_default; [] means "entry timeframe only". Validated
  # in trade_service. Each item is {"unit": "m"|"h"|"D"|"W", "value": <int>}.
  chart_timeframes: Optional[list[dict]] = None
  activities: list[TradeActivityCreate] = Field(min_length=1)
  tag_ids: list[int] = Field(default_factory=list)
  emotion_ids: list[int] = Field(default_factory=list)


class TradeUpdate(BaseModel):
  """Request body for updating a trade. All fields optional (partial update)."""

  model_config = ConfigDict(use_enum_values=True)

  asset_id: Optional[int] = None
  account_type: Optional[AccountType] = None
  stop_loss: Optional[float] = None
  notes: Optional[str] = Field(default=None, max_length=2000)
  missed_opportunity: Optional[bool] = None
  risk_per_trade: Optional[float] = None
  timeframe_unit: Optional[str] = Field(default=None, max_length=20)
  timeframe_value: Optional[int] = None
  # Per-trade chart timeframe overrides (issue #236). Omitted = leave unchanged
  # (exclude_unset); explicit null = reset to user defaults (store NULL);
  # [] = clear overrides and chart only the entry timeframe.
  chart_timeframes: Optional[list[dict]] = None
  activities: Optional[list[TradeActivityCreate]] = Field(default=None, min_length=1)
  tag_ids: Optional[list[int]] = None
  emotion_ids: Optional[list[int]] = None


class TradeSummary(BaseModel):
  """Lightweight trade representation for list responses — scalar columns only."""

  id: int
  asset_id: Optional[int] = None
  account_type: AccountType = AccountType.live
  status: TradeStatus
  direction: Optional[TradeDirection] = None
  stop_loss: Optional[float] = None
  notes: Optional[str] = None
  missed_opportunity: bool
  risk_per_trade: Optional[float] = None
  avg_entry_price: Optional[float] = None
  avg_exit_price: Optional[float] = None
  risk: Optional[float] = None
  reward: Optional[float] = None
  performance_r: Optional[float] = None
  timeframe_unit: Optional[str] = None
  timeframe_value: Optional[int] = None
  trade_date: Optional[str] = None
  created_at: Optional[str] = None
  updated_at: Optional[str] = None


class ResolvedTimeframe(BaseModel):
  """One effective chart timeframe for a trade (issue #236).

  `resolution` is the `{value}{unit}` token passed to the candles endpoint;
  `is_entry` flags the trade's entry timeframe within the resolved list.
  """

  unit: str
  value: int
  resolution: str
  is_entry: bool


class TradeResponse(TradeSummary):
  """Full trade detail, including nested activities, tags, emotions, screenshots."""

  # Raw per-trade override as stored (None = inherit user defaults), plus the
  # computed effective list the frontend renders (always present) — issue #236.
  chart_timeframes: Optional[list[dict]] = None
  resolved_chart_timeframes: list[ResolvedTimeframe] = Field(default_factory=list)
  activities: list[TradeActivityResponse] = Field(default_factory=list)
  tags: list[TagResponse] = Field(default_factory=list)
  emotions: list[EmotionResponse] = Field(default_factory=list)
  screenshots: list[TradeScreenshotResponse] = Field(default_factory=list)


class YearStatistics(BaseModel):
  """Aggregated statistics for all trades in a given year."""

  total_trades: int
  winning_trades: int
  losing_trades: int
  breakeven_trades: int
  total_pnl: float
  avg_pnl: float
  win_rate: float
  expectancy: float
  profit_factor: float

File: frontend/src/types/trade.ts
/**
 * Domain types for trades.
 *
 * Field names mirror the API's snake_case JSON so responses map directly
 * with no transform layer. `TradeSummary` matches the lightweight shape the
 * list endpoint (`GET /api/trades`) returns — scalar columns only.
 */

import type { Emotion, Tag } from '@/types/referenceData'

export type TradeStatus = 'Open' | 'Closed' | 'Partial' | 'Breakeven'

export type TradeDirection = 'Long' | 'Short'

export type AccountType = 'test' | 'demo' | 'live'

export interface TradeSummary {
  id: number
  asset_id: number | null
  account_type: AccountType
  status: TradeStatus
  direction: TradeDirection | null
  stop_loss: number | null
  notes: string | null
  missed_opportunity: boolean
  risk_per_trade: number | null
  avg_entry_price: number | null
  avg_exit_price: number | null
  risk: number | null
  reward: number | null
  performance_r: number | null
  timeframe_unit: string | null
  timeframe_value: number | null
  trade_date: string | null
  created_at: string | null
  updated_at: string | null
}

export type ActivityType = 'Buy' | 'Sell'

export interface TradeActivity {
  id: number
  trade_id: number
  type: ActivityType
  price: number
  quantity: number
  date: string
  is_entry: boolean
}

export interface TradeScreenshot {
  id: number
  trade_id: number
  filename: string
  timeframe_unit: string | null
  timeframe_value: number | null
  created_at: string | null
}

export interface TradeDetail extends TradeSummary {
  activities: TradeActivity[]
  tags: Tag[]
  emotions: Emotion[]
  screenshots: TradeScreenshot[]
}

/** A single buy/sell activity as sent to the create/update endpoints. */
export interface TradeActivityInput {
  type: ActivityType
  price: number
  quantity: number
  date: string
}

/**
 * Request body for creating or updating a trade. Mirrors the backend's
 * `TradeCreate` model; the update endpoint accepts the same shape.
 */
export interface TradeInput {
  asset_id: number
  account_type: AccountType
  stop_loss: number | null
  notes: string | null
  missed_opportunity: boolean
  risk_per_trade: number | null
  timeframe_unit: string | null
  timeframe_value: number | null
  activities: TradeActivityInput[]
  tag_ids: number[]
  emotion_ids: number[]
}
